core/src/main/java/org/apache/iceberg/BaseIncrementalAppendScan.java

```diff
@@ -0,0 +1,172 @@
+    List<Snapshot> snapshots = appendsBetween(table(), context().fromSnapshotId(), toSnapshotIdInclusive);
+    if (snapshots.isEmpty()) {
+      return CloseableIterable.empty();
+    }
+
+    Set<Long> snapshotIds = Sets.newHashSet(Iterables.transform(snapshots, Snapshot::snapshotId));
+    Set<ManifestFile> manifests = FluentIterable
+        .from(snapshots)
+        .transformAndConcat(Snapshot::dataManifests)
+        .filter(manifestFile -> snapshotIds.contains(manifestFile.snapshotId()))
+        .toSet();
+
+    ManifestGroup manifestGroup = new ManifestGroup(tableOps().io(), manifests)
+        .caseSensitive(context().caseSensitive())
+        .select(context().returnColumnStats() ? DataTableScan.SCAN_WITH_STATS_COLUMNS : DataTableScan.SCAN_COLUMNS)
+        .filterData(context().rowFilter())
+        .filterManifestEntries(
+            manifestEntry ->
+                snapshotIds.contains(manifestEntry.snapshotId()) &&
+                    manifestEntry.status() == ManifestEntry.Status.ADDED)
+        .specsById(tableOps().current().specsById())
+        .ignoreDeleted();
+
+    if (context().ignoreResiduals()) {
+      manifestGroup = manifestGroup.ignoreResiduals();
+    }
+
+    long fromSnapshotId = resolveFromSnapshotIdForListenerEvent(
+        table(), context().fromSnapshotId(), toSnapshotIdInclusive);
+    Listeners.notifyAll(new IncrementalScanEvent(table().name(), fromSnapshotId,
+        toSnapshotIdInclusive, context().rowFilter(), table().schema()));
+
+    if (manifests.size() > 1 &&
+        (DataTableScan.PLAN_SCANS_WITH_WORKER_POOL || context().planWithCustomizedExecutor())) {
+      manifestGroup = manifestGroup.planWith(context().planExecutor());
+    }
+
+    return manifestGroup.planFiles();
+  }
+
+  @Override
+  public CloseableIterable<CombinedScanTask> planTasks() {
+    CloseableIterable<FileScanTask> fileScanTasks = planFiles();
+    CloseableIterable<FileScanTask> splitFiles = TableScanUtil.splitFiles(fileScanTasks, targetSplitSize());
+    return TableScanUtil.planTasks(splitFiles, targetSplitSize(), splitLookback(), splitOpenFileCost());
+  }
+
+  /**
+   * Resolve nullable fromSnapshotIdExclusive to a primitive long type,
+   * as {@link IncrementalScanEvent} only accepts primitive long type.
+   */
+  private static long resolveFromSnapshotIdForListenerEvent(
+      Table table, Long fromSnapshotIdExclusive, long toSnapshotIdInclusive) {
+    if (fromSnapshotIdExclusive != null) {
+      return fromSnapshotIdExclusive;
+    } else {
+      // Resolve to the oldest ancestor of toSnapshotIdInclusive.
+      // This doesn't accurately capture the exclusive behavior of fromSnapshotId.
+      // Alternative is a special value (like 0L) to indicate null fromSnapshotIdExclusive.
+      Snapshot oldestAncestorSnapshot = SnapshotUtil.oldestAncestor(toSnapshotIdInclusive, table::snapshot);
+      return oldestAncestorSnapshot != null ? oldestAncestorSnapshot.snapshotId() : toSnapshotIdInclusive;
```

Comment:
Should we check that `oldestAncestorSnapshot.parentId() == fromSnapshotIdExclusive`?